apps/prairielearn/src/lib/workspace.js

```diff
@@ -46,615 +69,655 @@ const sql = sqldb.loadSqlEquiv(__filename);
+      `Deploying more computational resources (${t} seconds elapsed)`,
     );
+    await sleep(config.workspaceLaunchingRetryIntervalSec * 1000);
+    attempt++;
+  }
+  await workspaceUtils.updateWorkspaceMessage(workspace_id, 'Sending launch command to host');
+  await controlContainer(workspace_id, 'init', { useInitialZip });
+}
 
-    /** @type {WorkspaceFile[]} */
-    const dynamicFiles = (
-      await async.mapSeries(variant.params._workspace_files || [], async (file, i) => {
-        try {
-          // Ignore files without a name
-          if (!file.name) {
-            fileGenerationErrors.push({
-              file: `Dynamic file ${i}`,
-              msg: 'Dynamic workspace file does not include a name. File ignored.',
-              data: file,
-            });
-            return null;
-          }
-          // Discard names with directory traversal outside the home directory
-          if (!contains(remotePath, path.join(remotePath, file.name), false)) {
-            fileGenerationErrors.push({
-              file: file.name,
-              msg: 'Dynamic workspace file includes a name that traverses outside the home directory. File ignored.',
-              data: file,
-            });
-            return null;
-          }
+/**
+ * @typedef {Object} InitializeResult
+ * @property {string} sourcePath
+ * @property {string} destinationPath
+ */
 
-          if (file.questionFile) {
-            const localPath = path.join(questionBasePath, file.questionFile);
-            // Discard paths with directory traversal outside the question
-            if (!contains(questionBasePath, localPath, false)) {
-              fileGenerationErrors.push({
-                file: file.name,
-                msg: 'Dynamic workspace file points to a local file outside the question directory. File ignored.',
-                data: file,
-              });
-              return null;
-            }
-            // To avoid race conditions, no check if file exists here, rather an exception is
-            // captured when attempting to copy.
+/**
+ * This function constructs the initial state of a workspace.
+ *
+ * We'll construct the initial directory on disk in a temporary location and
+ * then return the path to that directory, as well as the path that the
+ * directory should be moved to. The caller is responsible for obtaining a
+ * lock for this workspace and moving the directory into its final location.
+ * Locking ensures that multiple hosts can't clobber writes to the same
+ * workspace. This is mostly important on NFS volumes, where renames (moves)
+ * are not atomic.
+ *
+ * @param {string | number} workspace_id
+ * @returns {Promise<InitializeResult>}
+ */
+export async function initialize(workspace_id) {
+  const { workspace, variant, question, course } = await sqldb.queryRow(
+    sql.select_workspace_data,
+    { workspace_id },
+    WorkspaceDataSchema,
+  );
+
+  // TODO Figure out what to do if course path is null
```

Comment:
Discussion led on Slack: course.path and question.qid are in essence not null, but the type is nullable. This need to be considered here.

Reply:
Added asserts with the assumption that workspaces without path/qid, if they exist at all, will not cause a workspace initialization.